File: sandbox.ts
import { formatEther, parseEther } from "viem";
import * as logger from "./logger";
import {getRandomDecimalInRange, truncateDecimal} from "./utils";
import * as fs from 'fs';


function updateHoldings(holdings: Map<string, bigint>, publicKey: string, amount: bigint): void {
    const currentHolding = holdings.get(publicKey) || BigInt(0);
    holdings.set(publicKey, currentHolding + amount);
}


export class Sandbox {
    private tokenSupply: bigint;
    private pooledEthLiquidity: bigint;
    private decimals: number;
    private tokenBalance: bigint;
    private ethBalance: bigint;
    public isRunning: boolean;
    public tokenHoldings: Map<string, bigint>;
    public ethHoldings: Map<string, bigint>;
    public state_snapshot: any;
    public initial_price: number;
    public ethBalanceSnap: bigint;
    public tokenBalanceSnap: bigint;

    constructor(tokenSupply: number, pooledEthLiquidity: number, decimals: number) {
        let tokenSupplyWei = parseEther(tokenSupply.toString());
        let pooledEthLiquidityWei = parseEther(pooledEthLiquidity.toString());

        this.tokenSupply = tokenSupplyWei;
        this.pooledEthLiquidity = pooledEthLiquidityWei;
        this.decimals = decimals;
        this.tokenBalance = tokenSupplyWei;
        this.ethBalance = pooledEthLiquidityWei;

        this.isRunning = false;
        this.tokenHoldings = new Map<string, bigint>();
        this.ethHoldings = new Map<string, bigint>();
        this.state_snapshot = null;

        this.initial_price = this.getPrice();
        this.ethBalanceSnap = this.ethBalance;
        this.tokenBalanceSnap = this.tokenBalance;
    }

    public buy(buyerPublicKey: string, ethInput: number, slippage: number): void {
        const tokenOutput = this.getTokenOutputForEthInput(ethInput);
        const ethWei = parseEther(ethInput.toString());
        const tokenOutputWei = parseEther(tokenOutput.toString());
        const delta_eth = this.getBalance(buyerPublicKey) - ethInput;
        
        this.tokenBalance -= tokenOutputWei;
        this.ethBalance += ethWei;

        updateHoldings(this.tokenHoldings, buyerPublicKey, tokenOutputWei);
        updateHoldings(this.ethHoldings, buyerPublicKey, -ethWei);

        const price = this.getPrice();

        const prevTokenBalance = this.tokenHoldings.get(buyerPublicKey) || BigInt(0);
        const prevEthBalance = this.ethHoldings.get(buyerPublicKey) || BigInt(0);

        const price_change = truncateDecimal((price - this.initial_price) / this.initial_price * 100, 2);
        logger.info(`BUY ${buyerPublicKey} ETH: ${ethInput} TOKEN: ${tokenOutput}, DELTA ETH: ${delta_eth}, PRICE CHANGE: ${price_change}%`)

    }

    public sell(sellerPublicKey: string, tokenInput: number, slippage: number): void {
        const minEthOutput = this.getEthOutputForTokenInput(tokenInput);
        const tokenInputWei = parseEther(tokenInput.toString());
        const ethOutputWei = parseEther(minEthOutput.toString());
        const token_delta = this.getTokenBalance(sellerPublicKey) - tokenInput;

        this.tokenBalance += tokenInputWei;
        this.ethBalance -= ethOutputWei;
        
        this.updateHoldings(this.tokenHoldings, sellerPublicKey, -tokenInputWei);
        this.updateHoldings(this.ethHoldings, sellerPublicKey, ethOutputWei);

        const price = this.getPrice();
        const price_change = truncateDecimal((price - this.initial_price) / this.initial_price * 100, 2);
        logger.info(`SELL ${sellerPublicKey} ETH: ${minEthOutput} TOKEN: ${tokenInput} DELTA TOKEN: ${token_delta}, PRICE CHANGE: ${price_change}%`)
        

    }

    public getPrice(): number {
        return Number(this.ethBalance) / Number(this.tokenBalance);
    }

    public getReserves(): { token0: number; token1: number } {
        return {
            token0: Number(formatEther(this.ethBalance)),
            token1: Number(formatEther(this.tokenBalance))
        };
    }

    public getBalance(publicKey: string): number {
        return Number(formatEther(this.ethHoldings.get(publicKey) || BigInt(0)));
    }

    public getTokenBalance(publicKey: string): number {
        return Number(formatEther(this.tokenHoldings.get(publicKey) || BigInt(0)));
    }

    public getTokenOutputForEthInput(ethInput: number): number {
        const ethInputWei = parseEther(ethInput.toString());
        const k = this.tokenBalance * this.ethBalance;
        return Number(formatEther(this.tokenBalance - (k / (this.ethBalance + ethInputWei))));
    }

    public getEthOutputForTokenInput(tokenInput: number): number {
        const tokenInputWei = parseEther(tokenInput.toString());
        const k = this.tokenBalance * this.ethBalance;
        return Number(formatEther(this.ethBalance - (k / (this.tokenBalance + tokenInputWei))));
    }

    private updateHoldings(holdings: Map<string, bigint>, publicKey: string, amount: bigint): void {
        const currentHolding = holdings.get(publicKey) || BigInt(0);
        holdings.set(publicKey, currentHolding + amount);
    }

    public distributeHoldings(
        minEth: number,
        maxEth: number,
        ratioWithoutTokens: number,
        wallets: string[]
    ): void {
        if (ratioWithoutTokens < 0 || ratioWithoutTokens > 1) {
            throw new Error("Ratio must be between 0 and 1");
        }

        const totalWallets = wallets.length;
        const walletsWithoutTokens = Math.round(totalWallets * ratioWithoutTokens);
        const walletsWithTokens = totalWallets - walletsWithoutTokens;
        logger.info(`walletsWithTokens: ${walletsWithTokens}`)
        logger.info(`walletsWithoutTokens: ${walletsWithoutTokens}`)

        for (let i = 0; i < totalWallets; i++) {
            const ethAmount = getRandomDecimalInRange(minEth, maxEth);
            this.updateHoldings(this.ethHoldings, wallets[i], parseEther(ethAmount.toString()));
        }

        const tokenPrice = Number(this.ethBalance) / Number(this.tokenBalance);
        logger.info(`Token price before distribution: ${tokenPrice}`);

        for (let i = 0; i < walletsWithTokens; i++) {
            const ethBalanceWei = this.ethHoldings.get(wallets[i]) || BigInt(0);
            const ethBalance = Number(formatEther(ethBalanceWei));
            const ethToBuy = ethBalance;

            this.buy(wallets[i], ethToBuy, 0);
        }
        const newTokenPrice = Number(this.ethBalance) / Number(this.tokenBalance);
        logger.info(`Token price after distribution: ${newTokenPrice}`);
        logger.info(`Token price change: ${(Number(newTokenPrice - tokenPrice) / Number(tokenPrice)) * 100}%`);

        this.initial_price = newTokenPrice;
        const balances = wallets.map(wallet => ({
            address: wallet,
            ethBalance: Number(formatEther(this.ethHoldings.get(wallet) || BigInt(0))),
            tokenBalance: Number(formatEther(this.tokenHoldings.get(wallet) || BigInt(0)))
        }));

        this.state_snapshot = balances;
        this.ethBalanceSnap = this.ethBalance;
        this.tokenBalanceSnap = this.tokenBalance;
    }

    public getGasPrice(): number {
        return 2.1;
    }

    public getAllWallets(): string[] {
        return Array.from(this.ethHoldings.keys());
    }
    public reloadState(): void {
        this.tokenBalance = this.tokenBalanceSnap;
        this.ethBalance = this.ethBalanceSnap;

        this.tokenHoldings = new Map<string, bigint>();
        this.ethHoldings = new Map<string, bigint>();

        for (const wallet of this.getAllWallets()) {
            this.tokenHoldings.set(wallet, this.state_snapshot.tokenHoldings.get(wallet) || BigInt(0));
            this.ethHoldings.set(wallet, this.state_snapshot.ethHoldings.get(wallet) || BigInt(0));
        }
    }

public getWalletChanges(): Map<string, { ethChange: number; tokenChange: number }> {
    const res = new Map<string, { ethChange: number; tokenChange: number }>();
    for (const wallet of this.state_snapshot) {
        const address = wallet.address;
        const initialEthBalance = wallet.ethBalance;
        const currentEthBalance = this.getBalance(address);
        const initialTokenBalance = wallet.tokenBalance;
        const currentTokenBalance = this.getTokenBalance(address);

        const ethChange = currentEthBalance - initialEthBalance
        const tokenChange = currentTokenBalance - initialTokenBalance
        res.set(address, { ethChange, tokenChange });
    }
    return res;

}
}

File: sandbox.py
import asyncio
from math import floor
from random import shuffle
import random
import time
from loguru import logger


class Sandbox:
    def __init__(self, token_supply, pooled_sol, decimals, fee=0.0001):
        self.token_supply = token_supply
        self.pooled_sol = pooled_sol
        self.decimals = decimals
        self.token_balance = token_supply
        self.sol_balance = pooled_sol
        self.is_running = False
        self.token_holdings = {}
        self.sol_holdings = {}
        self.fee = fee
        self.price_history = [{"price": self.get_price(), "timestamp": time.time()}]
        self.initial_price = self.get_price()
        self.snapshoot_after_distribution = {}
        self.tx = []

        with open("random_wallets_list.txt", "r") as f:
            self.wallets_for_random_tx = f.read().split("\n")
        
        asyncio.run(self.distribute_tokens(self.wallets_for_random_tx, 3, 0.5))
        


    def reset(self):
        self.token_holdings = {}
        self.sol_holdings = {}
        self.token_balance = self.token_supply
        self.sol_balance = self.pooled_sol
        self.price_history = [{"price": self.get_price(), "timestamp": time.time()}]
        self.initial_price = self.get_price()
        self.snapshoot_after_distribution = {}
        self.tx = []
        with open("random_wallets_list.txt", "r") as f:
            self.wallets_for_random_tx = f.read().split("\n")
        
        self.distribute_tokens(self.wallets_for_random_tx, 3, 0.5)


    async def buy(self, buyer_public_key, sol_input):
        token_output = self.get_token_output_for_sol_input(sol_input)

        self.token_balance -= token_output
        self.sol_balance += sol_input

        self.update_holdings(self.token_holdings, buyer_public_key, token_output)
        self.update_holdings(self.sol_holdings, buyer_public_key, -sol_input - self.fee)
        logger.info(f"BUY {buyer_public_key} {sol_input} {self.get_price()}")
        price = self.get_price()
        self.price_history.append({'price': price, 'timestamp': time.time()})
        self.tx.append({
            "type": "buy",
            "sender": buyer_public_key[-4:],
            "amount_in": sol_input,
            "amount_out": token_output,
            "price": price,
            "timestamp": time.time()
        })
        
    async def sell(self, seller_public_key, token_input):
        sol_output = self.get_sol_output_for_token_input(token_input)

        self.token_balance += token_input
        self.sol_balance -= sol_output

        self.update_holdings(self.token_holdings, seller_public_key, -token_input)
        self.update_holdings(self.sol_holdings, seller_public_key, sol_output-self.fee)
        logger.info(f"SELL {seller_public_key} {token_input} {self.get_price()}")
        price = self.get_price()
        self.price_history.append({'price': price, 'timestamp': time.time()})
        self.tx.append({
            "type": "sell",
            "sender": seller_public_key[-4:],
            "amount_in": token_input,
            "amount_out": sol_output,
            "price": price,
            "timestamp": time.time()
        })

    def get_price(self):
        return self.sol_balance / self.token_balance

    def get_balance(self, public_key):
        return self.sol_holdings.get(public_key, 0)

    def get_all_wallets(self):
        return [i for i in self.sol_holdings.keys() if i not in self.wallets_for_random_tx]
    def get_token_balance(self, public_key):
        return self.token_holdings.get(public_key, 0)

    def get_token_output_for_sol_input(self, sol_input):
        k = self.token_balance * self.sol_balance
        return self.token_balance - (k / (self.sol_balance + sol_input))

    def get_sol_output_for_token_input(self, token_input):
        k = self.token_balance * self.sol_balance
        return self.sol_balance - (k / (self.token_balance + token_input))

    def update_holdings(self, holdings, public_key, amount):
        current_holding = holdings.get(public_key, 0)
        holdings[public_key] = current_holding + amount

    async def distribute_tokens(self, wallets, sol_amount, holders_ratio):
        shuffle(wallets)

        for wallet in wallets:
            self.sol_holdings[wallet] = sol_amount

        holders_threshold = floor(len(wallets) * holders_ratio)
        for wallet in wallets[:holders_threshold]:
            await self.buy(wallet, (sol_amount-self.fee)*0.95)
        
        for wallet in wallets:
            self.snapshoot_after_distribution[wallet] = {
                "sol_balance": self.sol_holdings.get(wallet, 0),
                "token_balance": self.token_holdings.get(wallet, 0)
            }
    
    async def generate_random_transactions(self, num_txs, interval, regime):
        for _ in range(num_txs):
            while True:
                wallet = random.choice(self.wallets_for_random_tx)
                balance = self.get_balance(wallet)
                token_balance = self.get_token_balance(wallet)
                if regime == 'buy' and balance * 0.9 > 0.001:
                    amount = random.uniform(0.001, balance * 0.9)
                    await self.buy(wallet, amount)
                    break
                elif regime == 'sell' and token_balance > 0:
                    amount = random.uniform(1, token_balance)
                    await self.sell(wallet, amount)
                    break
                elif regime == 'shuffle':
                    type_ = random.choice(['buy', 'sell'])
                    if type_ == 'buy' and balance * 0.9 > 0.001:
                        amount = random.uniform(0.001, balance * 0.9)
                        await self.buy(wallet, amount)
                        break
                    elif type_ == 'sell' and token_balance > 0:
                        amount = random.uniform(1, token_balance)
                        await self.sell(wallet, amount)
                        break
            await asyncio.sleep(interval)
        
    
        